src/rdkafka_cgrp.c

```diff
@@ -3142,76 +3698,355 @@ rd_kafka_cgrp_max_poll_interval_check_tmr_cb (rd_kafka_timers_t *rkts,
+                        RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
+                        revoking, rd_true/*rejoin*/, "subscribe");
 
-        /* Trigger rebalance */
-        rd_kafka_cgrp_rebalance(rkcg, rd_true/*assignment lost*/,
-                                "max.poll.interval.ms exceeded");
+                rd_kafka_topic_partition_list_destroy(revoking);
+        }
+
+        if (unsubscribing_topics)
+                rd_kafka_topic_partition_list_destroy(unsubscribing_topics);
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
 }
 
 
 /**
  * Remove existing topic subscription.
  */
 static rd_kafka_resp_err_t
-rd_kafka_cgrp_unsubscribe (rd_kafka_cgrp_t *rkcg, int leave_group) {
-
-	rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "UNSUBSCRIBE",
-		     "Group \"%.*s\": unsubscribe from current %ssubscription "
-		     "of %d topics (leave group=%s, join state %s, v%"PRId32")",
-		     RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
-		     rkcg->rkcg_subscription ? "" : "unset ",
-		     rkcg->rkcg_subscription ? rkcg->rkcg_subscription->cnt : 0,
-		     leave_group ? "yes":"no",
-		     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state],
-		     rkcg->rkcg_version);
+rd_kafka_cgrp_unsubscribe (rd_kafka_cgrp_t *rkcg, rd_bool_t leave_group) {
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "UNSUBSCRIBE",
+                     "Group \"%.*s\": unsubscribe from current %ssubscription "
+                     "of size %d (leave group=%s, has joined=%s, %s, "
+                     "join-state %s)",
+                     RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                     rkcg->rkcg_subscription ? "" : "unset ",
+                     rkcg->rkcg_subscription ? rkcg->rkcg_subscription->cnt : 0,
+                     RD_STR_ToF(leave_group),
+                     RD_STR_ToF(RD_KAFKA_CGRP_HAS_JOINED(rkcg)),
+                     rkcg->rkcg_member_id ?
+                     rkcg->rkcg_member_id->str : "n/a",
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
 
         rd_kafka_timer_stop(&rkcg->rkcg_rk->rk_timers,
                             &rkcg->rkcg_max_poll_interval_tmr, 1/*lock*/);
 
-
         if (rkcg->rkcg_subscription) {
                 rd_kafka_topic_partition_list_destroy(rkcg->rkcg_subscription);
                 rkcg->rkcg_subscription = NULL;
         }
 
-	rd_kafka_cgrp_update_subscribed_topics(rkcg, NULL);
+        rd_kafka_cgrp_update_subscribed_topics(rkcg, NULL);
 
         /*
          * Clean-up group leader duties, if any.
          */
         rd_kafka_cgrp_group_leader_reset(rkcg, "unsubscribe");
 
-	if (leave_group)
-		rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_LEAVE_ON_UNASSIGN;
+        if (leave_group && RD_KAFKA_CGRP_HAS_JOINED(rkcg))
+                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_LEAVE_ON_UNASSIGN_DONE;
 
-        rd_kafka_cgrp_rebalance(rkcg, rd_false/*assignment not lost*/,
-                                "unsubscribe");
+        /* FIXME: Why are we only revoking if !assignment_lost ? */
```

Comment:
I think this is incorrect. I can't think why it's correct, and I don't remember thinking it through previously.

Reply:
i think before, if partitions were flagged as lost, this implied a revoke had already been triggered, so it's avoiding that.